src/renate/updaters/experimental/speft.py

```diff
@@ -0,0 +1,198 @@
+        self,
+        model: RenateModule,
+        loss_fn: torch.nn.Module,
+        optimizer: Callable[[List[Parameter]], Optimizer],
+        learning_rate_scheduler: Optional[Optional[Callable[[Optimizer], _LRScheduler]]] = None,
+        learning_rate_scheduler_interval: defaults.SUPPORTED_LR_SCHEDULER_INTERVAL_TYPE = defaults.LR_SCHEDULER_INTERVAL,  # noqa: E501
+        batch_size: int = defaults.BATCH_SIZE,
+        train_transform: Optional[Callable] = None,
+        train_target_transform: Optional[Callable] = None,
+        test_transform: Optional[Callable] = None,
+        test_target_transform: Optional[Callable] = None,
+        logged_metrics: Optional[Dict[str, torchmetrics.Metric]] = None,
+        seed: int = defaults.SEED,
+        mask_unused_classes: bool = defaults.MASK_UNUSED_CLASSES,
+    ) -> None:
+        if not isinstance(model, SPromptTransformer):
+            raise ValueError(
+                "SPrompt Learner can only be used with a SPromptTransformer model."
+                f"But got {type(model)}"
+            )
+        super().__init__(
+            model,
+            loss_fn,
+            optimizer,
+            learning_rate_scheduler,
+            learning_rate_scheduler_interval,
+            batch_size,
+            train_transform,
+            train_target_transform,
+            test_transform,
+            test_target_transform,
+            logged_metrics,
+            seed,
+            mask_unused_classes,
+        )
+
+    def on_model_update_start(
+        self,
+        train_dataset: Dataset,
+        val_dataset: Dataset,
+        train_dataset_collate_fn: Optional[Callable] = None,
+        val_dataset_collate_fn: Optional[Callable] = None,
+        task_id: Optional[str] = None,
+    ) -> None:
+        """A custom on_model_update_start hook for S-Peft methods.
+
+        Here, we iterate oer the train data set and extract features. These features used to compute
+        the task prototypes by the `update_task_identifier` call. Having this function in the model
+        update start instead of end results in val metrics being reflective of test accuracy.
+        """
+        super().on_model_update_start(
+            train_dataset, val_dataset, train_dataset_collate_fn, val_dataset_collate_fn, task_id
+        )
+        ## k-means
+        device = torch.device("cuda:0") if torch.cuda.is_available() else torch.device("cpu")
+        self._model.to(device)
+        features, labels = [], []
+        with torch.inference_mode():
+            for x, y in self.train_dataloader():
+                features.append(self._model.features(x.to(device)).cpu())
+                labels.append(y)
+        features = torch.cat(features)
+        labels = torch.cat(labels)
+        self._model.update_task_identifier(features=features, labels=labels)
+
+    def setup(self, stage: str) -> None:
+        # We dont support distributed
+        assert (
+            self.trainer.world_size == 1
```

Comment:
what's the problem with multi-gpu? some checkpoint issues?

Reply:
K-Means implementation is from sklearn. This endsup running num_gpus independent K-Means on
non-overalapping training sets (due to the distributed sampler; see line 106 above). Efforts to move
everything to `cuda:0` causes code-crashes that I didn't debug.